### ISSUE TITLE
third_party/repo.bzl file silently failed to apply patch

### ISSUE THREAD

**Issue by jerryyin** [tensorflow/tensorflow#30287]
<em>Please make sure that this is a build/installation issue. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:build_template</em>

**System information**
- OS Platform and Distribution: Linux Ubuntu 16.04:
- ~Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:~
- TensorFlow installed from (source or binary): source
- TensorFlow version: master branch
- Python version: 3
- ~Installed using virtualenv? pip? conda?:~
- Bazel version (if compiling from source): 0.24.1
- GCC/Compiler version (if compiling from source): 5.4.0
- ~CUDA/cuDNN version:~
- ~GPU model and memory:~

**Describe the problem**
[`git apply -v` in bazel scripts](https://github.com/tensorflow/tensorflow/blob/87f2f24f957abb76ad80e16b26b7221ddd277752/third_party/repo.bzl#L69) sliently failed to apply a patch. This cause tensorflow build to fail in an unexpected way.

The reason why `git apply -v` silently failed is due to [`git apply does not work from within the local checkout of an unrelated git repository`](https://stackoverflow.com/questions/24821431/git-apply-patch-fails-silently-no-errors-but-nothing-happens). In my setup, my home directory is a git directory used to sync up local development configuration files, and tensorflow build directory is under my home directory. This makes all tensorflow directories within the local checkout of a git directory (other than the one for project the patch is made for). Thus, at build time, the `_apply_patch()` subroutine will silently fail to apply any patch.

To fix, I will use the same patch command under either windows/linux to use the actual `patch -p1` command rather than `git apply`, ie, use [this line](https://github.com/tensorflow/tensorflow/blob/87f2f24f957abb76ad80e16b26b7221ddd277752/third_party/repo.bzl#L67) only. I didn't submit a PR because I'm not aware of why applying the patch need to branch on different OSes. If there is not a good reason to use `git apply`, I'd suggest to always use `patch -p1` instead.

~**Provide the exact sequence of commands / steps that you executed before running into the problem**~

~**Any other info / logs**
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached.~

**Comment by ravikyram** [comment]
Just to verify did you get chance to follow instructions from [TensorFlow ](https://www.tensorflow.org/install/source)website .Please, let us know. Thanks

**Comment by jerryyin** [comment]
@ravikyram Yes I do. Since official documentation never mentioned that it cannot be built from an existing git directory, it should not silently failing.

**Comment by mihaimaruseac** [comment]
I don't think this is a Tensorflow issue, seems to be caused by your special environment settings.

Git inside git is quite a no-no, to be honest, unless you're working with git submodules.

**Comment by jerryyin** [comment]
@mihaimaruseac While I agree it is very configuration related, what's the point of using different ways of apply patch on different OSes (knowing such a pitfall exists in `git apply`)?

BTW, at the point when `git apply` silently fails, it is not in the 

> Git inside git 

situation. It is in one level of git directory. It is due to bazel generated folders (symlink to `/root/.cache`) in a different git directory from where it is generated (TF).

**Comment by mihaimaruseac** [comment]
The windows special casing was added in 87f2f24f957abb76ad80e16b26b7221ddd277752 Before that change, all code used `git apply`, since 8ec93550847eaf9289ff33403bb19fa78c097334 which changed from `patch -p1` to `git apply` because `patch -p1` on CentOS doesn't support renaming of files.

Adding interested parties

**Comment by chsigg** [comment]
Sorry that this is causing trouble for you, and thanks Mihai for digging out the history and explaining the reason for the change. This was the best I could come up with to work around the limitations of `patch` on CentOS. Maybe there is a better way?

**Comment by jerryyin** [comment]
@mihaimaruseac @chsigg Thanks for explaining. Good to know the background knowledge on issue under CentOS. Given what the issue it is, I'd say the safest way to run `git apply` is to run it under the system root `/` where it is least likely to have git enabled (indicated by second point under [this answer](https://stackoverflow.com/a/27283285/3002734)).

I'm glad I'm able to figure it out, because it has choked me periodically for a while. But now I wouldn't mind if you guys leave the code as is, as it is caused by deficiencies of tooling in different platforms. Anyways I have automated my environment to get rid of git wrapping around bazel's generated files.

**Comment by meteorcloudy** [comment]
The patch command is always a headache on Windows because it's not a native tool. Looks like it is also a problem for Centos due to old versions. 
Bazel is implementing a native patch in Java, we'll expose an API `repository_ctx.patch` to allow users do patch in repository rules. It supports the most common features of patch command line tools (renaming, adding, deleting files), but will be more strict on the patch file content. This is still an ongoing work: https://github.com/bazelbuild/bazel/issues/8316.

**Comment by ravikyram** [comment]
@jerryyin  let me know if we can close this issue since it looks to be fixed. Thanks!

**Comment by jerryyin** [comment]
@ravikyram I knew the fix from beginning, but I would leave it open until the bazel implementation from @meteorcloudy is ready and replaced the logic here.

**Comment by mihaimaruseac** [comment]
I am +1 on leaving open until Bazel version is in production
